### PR TITLE
Add Unit test for testing smart group flow

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
Discussion PR on writing a unit test for the bug @MegaphoneJon is working on.

This also adds an exception into civiexit which was a test-game-changing suggestion from @totten (ping @seamuslee001 @monishdeb @mattwire @jitendrapurohit @pradpnayak  on that since it should mean we can handled civiexit now)

Next steps

1) agree exception naming & whether that part is right - I'll spin off the civiexit part into a separate PR.
2) @MegaphoneJon this test is incomplete but I think it solves the part of the problem you were struggling with - just work it into your own commit- (you don't need to keep it as a commit by me)
3) both my PR on the civiexit + exception & @MegaphoneJon should be against the rc as this relates to a regression

### CODE REVIEW

File: CRM/Core/Exception/PrematureExitException.php [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ *
+ * @param string $message
+ *   The human friendly error message.
+ * @param string $error_code
+ *   A computer friendly error code. By convention, no space (but underscore allowed).
+ *   ex: mandatory_missing, duplicate, invalid_format
+ * @param array $data
+ *   Extra params to return. eg an extra array of ids. It is not mandatory, but can help the computer using the api. Keep in mind the api consumer isn't to be trusted. eg. the database password is NOT a good extra data.
+ */
+class CRM_Core_Exception_PrematureExitException extends CRM_Core_Exception {
```
Comment (totten) [inline CRM/Core/Exception/PrematureExitException.php]: I'd lean toward [\RuntimeException](https://www.php.net/manual/en/class.runtimeexception.php) as a base-class.

Comment (totten) [reply CRM/Core/Exception/PrematureExitException.php]: Agree it's good to make a new class (e.g. `CRM_Core_Exception_PrematureExitException`) specifically for this situation. 👍

**Comment by eileenmcnaughton** [comment]
The new exception is merged & I think this can be merge too (once tests pass) - it adds marginal test coverage & some helpers to build on for the actual test & bug fix

**Comment by eileenmcnaughton** [comment]
@MegaphoneJon I'm merging this so you can build off it - it does add a little value just of itself

**Pull request merged**